### PR TITLE
Fix tick emit logic bugs in IB DataQueueHandler

### PR DESCRIPTION
**Pull request by StefanoRaggi**
#### Description
1. **Only emit quote ticks when both the price and size messages have been received**
Market data for trades and quotes are received by the IB API callbacks as separate messages, the price is always received before the size. Previously we were emitting ticks on both price and size messages.

2. **Do not emit half-quotes (if both quotes available)**
Even in very liquid instruments, we were previously sending quote ticks with e.g. only bid price and size but zero ask price and size. For consistency with other `IDataQueueHandler` implementations we now always emit complete quote ticks (for cases where there is no bid price available such as far OTM options, we'll emit the half-quote with zero for bid price and size).

3. **The tick Quantity field is now set only for Trade ticks**
Previously it was also being set for Quote ticks.

4. **Removed the possibility of emitting ticks with BidPrice >= AskPrice**
Previously this could happen.

#### Related Issue
n/a

#### Motivation and Context
Consistency and correctness of emitted ticks.

#### Requires Documentation Change
n/a

#### How Has This Been Tested?
- A live algorithm running locally with extensive logging of both IB messages received, ticks emitted and ticks/bars in `OnData` slices (for all security types, especially futures and options where this code is used in the cloud). 
- Manual inspection of the above logs has shown expected behavior.

#### Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by mchandschuh** [comment]
Regarding not emitting half quotes, is this desired? I'm recalling the raw data from when I was working on options and remember seeing many instances of having only a bid or ask update at a given time step. It seems like by preventing _half_ in all cases we may be filtering out potentially important data, **especially** in highly illiquid securities.

**Comment by StefanoRaggi** [comment]
> Regarding not emitting half quotes, is this desired? I'm recalling the raw data from when I was working on options and remember seeing many instances of having only a bid or ask update at a given time step. It seems like by preventing _half_ in all cases we may be filtering out potentially important data, **especially** in highly illiquid securities.

I probably wasn't clear enough in the PR description for that point: 
the actual bug being fixed is that when **both sides of the quote are available** only one half of it was being emitted at any time (one tick emitted for the bid with **zero ask** plus another tick emitted for the ask with **zero bid**).
In the case of an illiquid instrument with no bid or ask, the half quote **will** be emitted -- this case was tested with the example code in issue #3689.
In fact, when the very first tick for the subscription is received the half tick will always be emitted anyway (since we don't know if one of the sides is empty).

File: Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs [hunk 1]
```diff
@@ -2557,77 +2578,134 @@ private void HandleTickSize(object sender, IB.TickSizeEventArgs e)
         {
             Symbol symbol;
 
-            if (!_subscribedTickets.TryGetValue(e.TickerId, out symbol)) return;
+            if (!_subscribedTickets.TryGetValue(e.TickerId, out symbol))
```
Comment (Martin-Molinero) [inline Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs]: An idea: Maybe `_subscribedTickets` could hold the `Symbol` and all `Tick` types in a new class, would help us avoid the extra `ConcurrentDictionaries`, their access and the `if(!) add` logics, because since we are not removing them, after the first tick the symbol will already be in the dictionaries ?

File: Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs [hunk 2]
```diff
@@ -2557,77 +2578,134 @@ private void HandleTickSize(object sender, IB.TickSizeEventArgs e)
...
-                    tick.BidSize = tick.Quantity;
+                    if (tick.BidPrice == 0)
+                    {
+                        // no bid price, do not emit tick
+                        return;
+                    }
+
+                    if (tick.BidPrice > 0 && tick.AskPrice > 0 && tick.BidPrice >= tick.AskPrice)
+                    {
+                        // new bid price jumped above previous ask price, wait for new ask price
```
Comment (Martin-Molinero) [inline Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs]: Another idea 🤔  would it be worth emitting half a bid quote here, setting the ask values to 0?

File: Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs [hunk 3]
```diff
@@ -2557,77 +2578,134 @@ private void HandleTickSize(object sender, IB.TickSizeEventArgs e)
...
-            var tick = new Tick();
-            // in the event of a symbol change this will break since we'll be assigning the
-            // new symbol to the permtick which won't be known by the algorithm
-            tick.Symbol = symbol;
             var securityType = symbol.ID.SecurityType;
-            tick.Quantity = AdjustQuantity(securityType, e.Size);
-            tick.Time = GetRealTimeTickTime(symbol);
-
-            if (tick.Quantity == 0) return;
+            var quantity = AdjustQuantity(securityType, e.Size);
```
Comment (Martin-Molinero) [inline Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs]: I think `quantity` could also be reported as `-1` in some cases, should we add a check for it? `"...When IBApi::EWrapper::tickPrice and IBApi::EWrapper::tickSize are reported as -1, this indicates that there is no data currently available...."`

File: Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs [hunk 4]
```diff
@@ -2481,62 +2481,83 @@ private DateTime GetRealTimeTickTime(Symbol symbol)
...
+                            // in the event of a symbol change this will break since we'll be assigning the
+                            // new symbol to the permtick which won't be known by the algorithm
+                            Symbol = symbol,
+                            TickType = TickType.Quote
+                        };
+                        _lastQuoteTicks.TryAdd(symbol, tick);
+                    }
+
+                    // negative price (-1) means no price available, normalize to zero
+                    tick.BidPrice = price < 0 ? 0 : price;
```
Comment (Martin-Molinero) [inline Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs]: Maybe we could normalize the price outside the switch so it's shared? (doesn't it also apply for `IBApi.TickType.LAST`?

Comment (StefanoRaggi) [reply Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs]: I've never seen a negative price for `IBApi.TickType.LAST` but will move the normalization up to remove duplication

File: Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs (hunk 1 above)
Comment (StefanoRaggi) [reply Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs]: Good idea, will do 👍

File: Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs (hunk 3 above)
Comment (StefanoRaggi) [reply Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs]: Do you have a link for that quote?

Looking at the latest IB API docs for `tickPrice` here:
https://interactivebrokers.github.io/tws-api/interfaceIBApi_1_1EWrapper.html#ae851ec3a1e0fa2d0964c7779b0c89718
```
Market data tick price callback. Handles all price related ticks. 
Every tickPrice callback is followed by a tickSize. 
A tickPrice value of -1 or 0 followed by a tickSize of 0 indicates there is no data for this field currently available, whereas a tickPrice with a positive tickSize indicates an active quote of 0 (typically for a combo contract).
```
This was also confirmed by my logs during testing, where I have never seen a negative `tickSize` returned.

File: Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs (hunk 2 above)
Comment (StefanoRaggi) [reply Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs]: Hmm... point 2 of this PR was to actually prevent us from emitting incomplete/half ticks when possible 😃 

In these normal bid/ask price changes, i.e. when both bid and ask quotes move up or down **and** are both non-zero, we will always receive both half-quotes in sequence.

File: Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs (hunk 3 above)
Comment (Martin-Molinero) [reply Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs]: Right, link @ https://interactivebrokers.github.io/tws-api/md_receive.html#market_data_return 3rd paragraph, I think its the case where there is no data available at all vs the field, I guess adding the check for -1 won't cause any harm?

Comment (StefanoRaggi) [reply Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs]: Ok, I'll add the normalization for negative quantity as well.
The Trade tick with non-positive quantity will be discarded anyway by `Tick.IsValid()`.
